### PR TITLE
Development pull request

### PR DESCRIPTION
**Pull request by felipebaltazar**

### CODE REVIEW

File: InfoMovies/InfoMovies/Converters/AspectRatioConverter.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        #region Properties
+
+        public float Ratio { set; get; }
+
+        #endregion
+
+        #region IValueConverter
+
+        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
```
Comment (arctouch-joseclua) [inline InfoMovies/InfoMovies/Converters/AspectRatioConverter.cs]: How could you improve this Converter so it becomes more secure? (Less error prone)

File: InfoMovies/InfoMovies/Mappings/Mapping.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
+using System.Linq.Expressions;
+
+namespace InfoMovies.Mappings
+{
+    public abstract class Mapping<TSource, TDestination>
```
Comment (arctouch-joseclua) [inline InfoMovies/InfoMovies/Mappings/Mapping.cs]: Have you worked with MVVM frameworks such as Prism or MVVMCross? Do you prefer this Model / ViewModel mapping solution?  If so, why?

File: InfoMovies/InfoMovies/Views/ItemDetailsPage.xaml [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                            <ColumnDefinition Width="70*"/>
+                        </Grid.ColumnDefinitions>
+
+                        <Views:GradientMask TopColor="Black" BottomColor="Gray" 
+                                        Grid.Row="0" Grid.Column="0" Grid.RowSpan="3" Grid.ColumnSpan="2"/>
+
+                        <ImageLoading:CachedImage HorizontalOptions="FillAndExpand" VerticalOptions="FillAndExpand"
+                                                  Grid.Row="0" Grid.RowSpan="3" Grid.Column="0" Aspect="AspectFit" Source="{Binding Path=Poster}"/>
+
+                        <StackLayout Orientation="Vertical" Style="{StaticResource stackLayoutStyle}"
```
Comment (arctouch-joseclua) [inline InfoMovies/InfoMovies/Views/ItemDetailsPage.xaml]: Do you really need this StackLayout? If so, why?

File: InfoMovies/InfoMovies/Views/ItemDetailsPage.xaml [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+                                            <ImageLoading:CachedImage HorizontalOptions="FillAndExpand" VerticalOptions="FillAndExpand"
+                                                                Aspect="AspectFit" Source="{Binding Path=ProfilePath}"/>
+                                        </Grid>
+                                        
+                                        <StackLayout Orientation="Horizontal" Padding="0" Style="{StaticResource stackLayoutStyle}">
+                                            <Label Style="{StaticResource labelStyle}" Text="{Extensions:Translate Text=ItemDetailsPage_Name}"
+                                                   LineBreakMode="NoWrap" TextColor="Black" VerticalTextAlignment="Center"
+                                                   HorizontalOptions="Start" FontSize="Micro"/>
+                                            <Label Style="{StaticResource labelStyle}" Text="{Binding Name}" LineBreakMode="NoWrap"
+                                                   TextColor="Black" VerticalTextAlignment="Center" HorizontalOptions="End" FontSize="Micro"/>
```
Comment (arctouch-joseclua) [inline InfoMovies/InfoMovies/Views/ItemDetailsPage.xaml]: How could you improve the use of Styles, for example lableStyle?

File: InfoMovies/InfoMovies/Views/ItemDetailsPage.xaml.cs [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+using Xamarin.Forms.Xaml;
+
+namespace InfoMovies.Views
+{
+    [XamlCompilation(XamlCompilationOptions.Compile)]
+	public partial class ItemDetailsPage : BasePage
+	{
+		public ItemDetailsPage (object viewModel)
+		{
+            BindingContext = viewModel;
```
Comment (arctouch-joseclua) [inline InfoMovies/InfoMovies/Views/ItemDetailsPage.xaml.cs]: How can we improve this?

File: InfoMovies/InfoMovies/Views/MainPage.xaml [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+                    <RowDefinition Height="50"/>
+                    <RowDefinition Height="Auto"/>
+                </Grid.RowDefinitions>
+                
+                <Grid.ColumnDefinitions>
+                    <ColumnDefinition Width="*"/>
+                    <ColumnDefinition Width="80"/>
+                </Grid.ColumnDefinitions>
+                
+                <Label Text="{Extensions:Translate Text=MainPage_NoData}"
```
Comment (arctouch-joseclua) [inline InfoMovies/InfoMovies/Views/MainPage.xaml]: How is the NoData text going to be hidden when list is not empty?

File: InfoMovies/InfoMovies/Views/BasePage.xaml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+﻿<?xml version="1.0" encoding="utf-8" ?>
+<ContentPage xmlns="http://xamarin.com/schemas/2014/forms"
+             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
+             x:Class="InfoMovies.Views.BasePage"
+             NavigationPage.HasNavigationBar="false"
```
Comment (arctouch-joseclua) [inline InfoMovies/InfoMovies/Views/BasePage.xaml]: How will the user navigate back from item details to the main page in iOS ?

File: InfoMovies/InfoMovies/Converters/AspectRatioConverter.cs (hunk 1 above)
Comment (felipebaltazar) [reply InfoMovies/InfoMovies/Converters/AspectRatioConverter.cs]: validating the `Ratio` parameter and creating exceptions to notify developer (like `DivideByZeroException` and `NullParameter`)

File: InfoMovies/InfoMovies/Mappings/Mapping.cs (hunk 2 above)
Comment (felipebaltazar) [reply InfoMovies/InfoMovies/Mappings/Mapping.cs]: I never worked with MVVM frameworks... :(
i needed to "re-discover wheel" when i learning about xamarin ... but this case makes me better developer...
I always willing to learn new technologies, frameworks, patterns and etc... :)

File: InfoMovies/InfoMovies/Views/ItemDetailsPage.xaml (hunk 3 above)
Comment (felipebaltazar) [reply InfoMovies/InfoMovies/Views/ItemDetailsPage.xaml]: no really...
my bad

File: InfoMovies/InfoMovies/Views/ItemDetailsPage.xaml (hunk 4 above)
Comment (felipebaltazar) [reply InfoMovies/InfoMovies/Views/ItemDetailsPage.xaml]: creating inherited styles according to the repetition of controls with similar appearance
I have created several labels repeating some properties. :(

File: InfoMovies/InfoMovies/Views/BasePage.xaml (hunk 7 above)
Comment (felipebaltazar) [reply InfoMovies/InfoMovies/Views/BasePage.xaml]: i'm really newbie at IOS development..
So, i dont tested with this plataform and for that reason i forget the creation of back buton

Im  thinkig about `swipe gesture` or 'back float button' to bring it back 
(in this project i didnt not want to use navigationbar for design of app)

File: InfoMovies/InfoMovies/Views/ItemDetailsPage.xaml.cs (hunk 5 above)
Comment (felipebaltazar) [reply InfoMovies/InfoMovies/Views/ItemDetailsPage.xaml.cs]: now, i dont can think a way different that:

`new ItemDetailPage(){ BindingContext = viewModel};`

I searched about another ways to do this, but dont find a "better way" for this
any suggest?  😆

File: InfoMovies/InfoMovies/ViewModels/MainPageViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+                if (!isSearching)
+                {
+                    currentPage = 1;
+                    movies.Clear();
+                }
+
+                isSearching = true;
+
+                var tmdbMovies =  await _provider.SearchAsync(searchText, currentPage);
+                FillMoviesCollection(tmdbMovies);
```
Comment (arctouch-sidneipolo) [inline InfoMovies/InfoMovies/ViewModels/MainPageViewModel.cs]: What if the tmdbMovies list is null? As you have handled to return null every time that something goes wrong, you should care about handling null objects when making requests. How can you improve the error handling especially for web requests?

File: InfoMovies/InfoMovies/Views/BasePage.xaml.cs [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+using Xamarin.Forms;
+using Xamarin.Forms.Xaml;
+
+namespace InfoMovies.Views
+{
+	[XamlCompilation(XamlCompilationOptions.Compile)]
```
Comment (arctouch-sidneipolo) [inline InfoMovies/InfoMovies/Views/BasePage.xaml.cs]: Why is this line needed?

File: InfoMovies/InfoMovies/Views/MainPage.xaml.cs [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        }
+
+        #endregion
+
+        #region Private Methods
+
+        private async void InitializeItemsAsync()=>
+            await viewModel.LoadItemsAsync();
+
+        private async void ListView_ItemAppearing(object sender, ItemVisibilityEventArgs e)
```
Comment (arctouch-sidneipolo) [inline InfoMovies/InfoMovies/Views/MainPage.xaml.cs]: Let's say that we need to use a infinity list in more than one page. How can you improve it so it can be reused?

Comment (felipebaltazar) [reply InfoMovies/InfoMovies/Views/MainPage.xaml.cs]: We could create a custom control, just like the horizontal list.
And to make it easier, it would create an asynchronous collection with the "MoveNext" and "FillList" commands to be implemented

File: InfoMovies/InfoMovies/ViewModels/MainPageViewModel.cs (hunk 8 above)
Comment (felipebaltazar) [reply InfoMovies/InfoMovies/ViewModels/MainPageViewModel.cs]: i skipped that, my bad... D:
We can replace `null` returns to  `Default()`

File: InfoMovies/InfoMovies/Views/BasePage.xaml.cs (hunk 9 above)
Comment (felipebaltazar) [reply InfoMovies/InfoMovies/Views/BasePage.xaml.cs]: to check Xaml errors at compile time and improve application load.

File: InfoMovies/InfoMovies/ViewModels/MainPageViewModel.cs (hunk 8 above)
Comment (arctouch-sidneipolo) [reply InfoMovies/InfoMovies/ViewModels/MainPageViewModel.cs]: Returning default() here would cause the same problem. The default value for a list is null.

File: InfoMovies/InfoMovies/Views/BasePage.xaml.cs (hunk 9 above)
Comment (arctouch-sidneipolo) [reply InfoMovies/InfoMovies/Views/BasePage.xaml.cs]: Sure, but you’ve added it at application  level. There’s no need to add it again in every page.

File: InfoMovies/InfoMovies/ViewModels/MainPageViewModel.cs (hunk 8 above)
Comment (felipebaltazar) [reply InfoMovies/InfoMovies/ViewModels/MainPageViewModel.cs]: yeah! thats true.. sorry, i don't make me clear..
i was thinking about web requests (generics), to return default(T)
Like `Web Api Client`
But talking about collections x webrequest, we can crate a default collection and fill that with return of webrequest, if not null.
So, we can return this collection with web request return or a collection with none items

File: InfoMovies/InfoMovies/Views/BasePage.xaml.cs (hunk 9 above)
Comment (felipebaltazar) [reply InfoMovies/InfoMovies/Views/BasePage.xaml.cs]: Really? I didn't know!
Thanks for that!
